openssl can fail, leaving an empty file and causing a reboot loop

yanokwa opened the pull request:
openssl can fail, leaving an empty file and causing a reboot loop since `/etc/dh/nginx.pem` cannot be found.

During an upgrade to Central v1.5.3, this is what the user saw when the server did not come up.
![image](https://user-images.githubusercontent.com/32369/192063719-e17129a4-f7d3-415e-97dc-f8960781911d.png)

Logs showed it was an issue with `/etc/dh/nginx.pem`.
![image](https://user-images.githubusercontent.com/32369/192063692-d5526113-ce79-45e0-aee0-18481a902aa5.png)

Not sure what happened, but it seems that for whatever reason, the openssl command failed and resulted in an empty file.

This PR checks for an empty file in all places where openssl is called. I confirmed the syntax is correct with shellcheck. I also wrap the mkdir command in case we ever use a white space in the name.

I confirmed this fix works on the server that had this issue.

matthew-white commented:
@alxndrsn, if you have time, would you mind taking a look?

matthew-white commented:
I'm going to target the `release` branch with the thought that this change would go out with v2022.3.

alxndrsn commented on files/nginx/odk-setup.sh:
```diff
@@ -1,12 +1,12 @@
 DHPATH=/etc/dh/nginx.pem
-if [ ! -e "$DHPATH" ] && [ "$SSL_TYPE" != "upstream" ]
+if [[ (! -e "$DHPATH" || ! -s "$DHPATH" ) && "$SSL_TYPE" != "upstream" ]]
```
is `-e` made redundant by `-s`?

From the bash manpage:
```
       -s file
              True if file exists and has a size greater than zero.
```

alxndrsn commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,13 @@
+
+#### Before submitting this PR, please make sure you have:
+
+- [ ] verified that any code or assets from external sources are properly credited in comments.
```
Changes seem unrelated to the PR title.

alxndrsn commented on docs/news.html:
```diff
@@ -8,15 +8,15 @@
-        ODK Central v1.5.2
+      <span class="item-date">2022/09/15</span>
+      <a href="https://forum.getodk.org/t/backups-to-google-drive-from-central-will-stop-working-after-jan-31st/38895" target="_blank">
+        ⚠️ Google backups sunsetted ⚠️
```
Changes seem unrelated to the PR title.

alxndrsn reviewed (commented):
> I confirmed the syntax is correct with shellcheck. I also wrap the mkdir command in case we ever use a white space in the name.

I get two shellcheck violations:

```
In odk-setup.sh line 1:
DHPATH=/etc/dh/nginx.pem
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.


In odk-setup.sh line 12:
  mkdir -p $SELFSIGN_BASEPATH
           ^----------------^ SC2086: Double quote to prevent globbing and word splitting.
```

Is there a reason not to use bash's `-e` and `-o pipefail` options?  This might give more robust handling of unforeseen failures of individual commands in this file.

yanokwa commented:
Sorry about the extra commits, I've cleaned that up and also fixed the redundant flag. The script should pass shellcheck.

As to `-e` and `-o pipefail`, maybe those can come as part of https://github.com/getodk/central/pull/328 so we can do that work at once on all scripts.

yanokwa commented:
I went for the narrow change.

I don't know offhand how to fail immediately here. In what I've seen, the container will restart endlessly when something goes wrong.
